tests/test_datacollection.py

```diff
@@ -0,0 +1,197 @@
+        # duplicate keys
+        dc["label 1"] = data_object_1_b
+
+
+def test_loaded_derived_data(create_data_object: Callable):
+    data_object_loaded_1 = create_data_object("label 1")
+    data_object_loaded_2 = create_data_object("label 2")
+    data_object_derived_1_a = create_data_object("label 1 der a", derived_from=[data_object_loaded_1])
+    data_object_derived_1_b = create_data_object("label 1 der b", derived_from=[data_object_loaded_1])
+    data_object_derived_2_a = create_data_object("label 2 der a", derived_from=[data_object_loaded_2])
+
+    dc = DataCollection(ContinuousData)
+    dc.add(
+        data_object_loaded_1,
+        data_object_loaded_2,
+        data_object_derived_1_a,
+        data_object_derived_1_b,
+        data_object_derived_2_a,
+    )
+
+    assert dc.get_loaded_data() == {
+        "label 1": data_object_loaded_1,
+        "label 2": data_object_loaded_2,
+    }
+    assert dc.get_derived_data() == {
+        "label 1 der a": data_object_derived_1_a,
+        "label 1 der b": data_object_derived_1_b,
+        "label 2 der a": data_object_derived_2_a,
+    }
+    assert dc.get_data_derived_from(data_object_loaded_1) == {
+        "label 1 der a": data_object_derived_1_a,
+        "label 1 der b": data_object_derived_1_b,
+    }
+    assert dc.get_data_derived_from(data_object_loaded_2) == {
+        "label 2 der a": data_object_derived_2_a,
+    }
+
+
+def test_concatenate(create_data_object: Callable):
+    time = np.arange(100) / 20
+    rng = np.random.default_rng()
+    values = rng.random(time.shape)
+
+    data_object_1_a = create_data_object("label 1", time=time[:30], values=values[:30])
+    data_object_1_b = create_data_object("label 1", time=time[30:], values=values[30:])
+    data_object_2 = create_data_object("label 2", time=time[30:], values=values[30:])
+
+    dc_1_a = DataCollection(ContinuousData)
+    dc_1_b = DataCollection(ContinuousData)
+    dc_2 = DataCollection(ContinuousData)
+
+    dc_1_a.add(data_object_1_a)
+    dc_1_b.add(data_object_1_b)
+    dc_2.add(data_object_2)
+
+    dc_1_concat = dc_1_a.concatenate(dc_1_b)
+    assert np.array_equal(dc_1_concat["label 1"].time, time)
+    assert np.array_equal(dc_1_concat["label 1"].values, values)
+
+    with pytest.raises(ValueError, match="Keys don't match"):
+        dc_1_a.concatenate(dc_2)
+
+    with pytest.raises(ValueError, match=r"(.*?) \(b\) starts before (.*?) \(a\) end"):
+        dc_1_b.concatenate(dc_1_a)
+
+
+def test_select_by_time(create_data_object: Callable):
+    time1 = np.arange(0, 100)
+    time2 = np.arange(50, 150)
+    rng = np.random.default_rng()
+    values1 = rng.random(time1.shape)
+    values2 = rng.random(time2.shape)
+
+    data_object_1 = create_data_object("label 1", time=time1, values=values1)
+    data_object_2 = create_data_object("label 2", time=time2, values=values2)
+
+    dc = DataCollection(ContinuousData)
+    dc.add(data_object_1, data_object_2)
+
+    s1 = dc.select_by_time(20, 80)
```

Comment:
Perhaps also test what happens if start and end of select_by_time are the same?

Reply:
This is a good suggestion. However, `select_by_time` currently is not in a great state and will be rewritten partly.
Therefore, I'll add is as TODO.